mlflow/sklearn/__init__.py

```diff
@@ -455,6 +461,271 @@ def load_model(model_uri):
+    (2) a map of "prediction result object id" to a tuple of dataset name(the dataset is
+       the one which generate the prediction result) and model id.
+    (3) some scope flags which indicates currently in which method scope, we record scope
+       flags for method `model.fit`, `eval_and_log_metrics`, `model.score`.
+       In order to avoid nested/duplicated autologging metric, when run into these scopes,
+       we need temporarily disable the metric autologging.
+    (4) eval_dataset_info_map, it is a double level map:
+       `eval_dataset_info_map[model_uuid][eval_dataset_var_name]` will get a list, each
+       element in the list is a tuple of `(id(eval_dataset_instance), eval_dataset_sample_rows)`
+       This data structure is used for:
+        * generating unique dataset name key when autologging metric
+        * storing the dataset row samples, we need log it into metric_info artifact file.
+    (5) metric_api_call_arg_dict_list_map, it is a map, key is metric api function name,
+       and value is a list of call arguments dict.
+    """
+
+    fit_scope_key = 'fit'
+    eval_and_log_metrics_scope_key = 'eval_and_log_metrics'
+    model_score_scope_key = 'model_score'
+
+    def __init__(self):
+        self.model_id_to_run_id_map = {}
+        self.pred_result_id_to_dataset_name_and_model_id = {}
+        self.scope_flag_dict = {
+            self.fit_scope_key: False,
+            self.eval_and_log_metrics_scope_key: False,
+            self.model_score_scope_key: False
+        }
+        self.eval_dataset_info_map = defaultdict(lambda: defaultdict(list))
+        self.metric_api_call_arg_dict_list_map = defaultdict(list)
+
+    def call_scope(self, scope_key):
+        class CallScope:
+            def __enter__(inner_self):
+                self.scope_flag_dict[scope_key] = True
+
+            def __exit__(inner_self, exc_type, exc_val, exc_tb):
+                self.scope_flag_dict[scope_key] = False
+
+        return CallScope()
+
+    def get_run_id_for_model(self, model):
+        return self.model_id_to_run_id_map.get(model._mlflow_uuid, None)
+
+    def should_log_eval_metrics(self, metric_value=None):
+        """
+        Test whether we should log metric. will check metric value types (if provided)
+        and scope flags
+        """
+        return (np.isscalar(metric_value) if metric_value is not None else True) and \
+            not self.scope_flag_dict[self.fit_scope_key] and \
+            not self.scope_flag_dict[self.eval_and_log_metrics_scope_key] and \
+            not self.scope_flag_dict[self.model_score_scope_key]
+
+    def register_model(self, model, run_id):
+        """
+        In `patched_fit`, we need register the model with the run_id used in `patched_fit`
+        So that in following metric autologging, the metric will be logged into the registered
+        run_id
+        """
+        model._mlflow_uuid = uuid4()
+        self.model_id_to_run_id_map[model._mlflow_uuid] = run_id
```

Comment:
Do we need the map, or can we just assign the `run_id` to the `model` as an attribute? 

E.g. `model._run_id = run_Id`?

Reply:
In `patched_metric_api`, we don't have model context, we can only query out the "model_id" by the `id(argument)`, then we need query "run_id" via "model_id",  if we put "run_id" inside model instance, then we have to create another map of: model_id --> model_instance

So this way cannot reduce complexity.